keps/sig-network/0031-service-topology.md

```diff
@@ -0,0 +1,168 @@
+
+### New PodLocator resource
+
+As EndpointAddress already contains nodeName field, we can build a service that will precook Pod to its topologies mapping. Then let all interested components(at least kube-proxy and kube-dns and coredns) just watch that precooked object and do necessary mapping internally. Given that we don't know which labels are topology labels, we are going to copy all node labels.
+
+```
+// PodLocator represents information about where a pod exists in arbitrary space.  This is useful for things like
+// being able to reverse-map pod IPs to topology labels, without needing to watch all Pods or all Nodes.
+type PodLocator struct {
+    metav1.TypeMeta
+    // +optional
+    metav1.ObjectMeta
+
+    // NOTE: Fields in this resource must be relatively small and relatively low-churn.
+
+    IPs []PodIPInfo // being added for dual-stack support
+    NodeName string
+    NodeLabels map[string]string
+}
+```
+
+In order to reference PodLocator back to a Pod easily, PodLocator namespace and name would be 1:1 with Pod namespace and name. In other words, PodLocator is a lightweight object which stores Pod location/topology information
+
+### New PodLocator controller
+
+A new PodLocator controller will watch and cache all Pods and Nodes. Then pre-cook pod name to {pod IPs, node name, node labels} mapping. 
+
+When a Pod is added, PodLocator controller will created a new PodLocator object whose namespace and name are 1:1 with Pod namespace and name. Then it will pupolate Pod's IP(s), node name and labels into the new object.
+
+When a Pod is updated, PodLocator controller will first check if IPs or Spec.NodeName are changed. If changed, PodLocator controller will update the corresponding PodLocator object accordingly, otherwise will ignore this change.
+
+When a Pod is deleted, PodLocator controller will delete the corresponding PodLocator object.
+
+When a Node is updated, PodLocator controller will first check if its labels are changed. If changed, will update all the PodLocators whose corresponding Pods running on it.
+
+When a Node is deleted, PodLocator controller will reset the NodeName and NodeLabels of all the PodLocators whose corresponding Pods running on it.
+
+### Kube-proxy changes
+
+Kube-proxy will respect topology keys for each service, so kube-proxy on different nodes may create different proxy rules.
+
+Kube-proxy will watch its own node and will find the endpoints that are in the same topological domain as the node if `service.TopologyKeys` is not empty.
+
+Kube-proxy will watch PodLocator apart from Service and Endpoints. For each Endpoints object, kube-proxy will find the original Pod via EndpointAddress.TargetRef, therefore will get PodLocator object and its topology information. Kube-proxy will only create proxy rules for endpoints that are in the same topological domain as the node running kube-proxy.
```

Comment:
Why does kube-proxy need PodLocator?  It already knows its own Node.

Reply:
I think kube-proxy still needs PodLocator as it need to map from `Endpoints.subsets[].addresses.nodeName` -> `Node.metadata.labels`.